### PR TITLE
add a disk_free_space check before writing config

### LINKED ISSUE #25175: Config file truncated when low on disk space
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### The bug

When the disk that Nextcloud is installed on is full (an admittedly unpredictable scenario), the config file is truncated. This happened during an update in my case.

### Steps to reproduce
1. `cat config/config.php`, observe config values
2. Fill up disk space, e.g. `dd if=/dev/zero of=file.dat bs=1024 count=10000000`
2. Perform action that changes config file, e.g. `occ maintenance:mode --on`
3. `cat config/config.php` observe an empty file

### Expected behaviour
Config file is not truncated when no disk space.

### Actual behaviour
Config is left empty, Nextcloud believes it is no longer installed and wants a CAN_INSTALL file.

### Server configuration

**Operating system:**
Ubuntu 18.04.5 LTS

**Web server:**
apache2 2.4.29-1ubuntu4.14 

**Database:**
mariadb 10.1.44

**PHP version:**
PHP 7.2.24-0ubuntu0.18.04.6

**Nextcloud version:** (see Nextcloud admin page)
20.0.5

### PR DESCRIPTION
**Pull request by szaimen**
Fix https://github.com/nextcloud/server/issues/25175

Idea: the config file should simply not get written if disk space is low. Instead the current process should be exited. This should hopefully fix such issues in the future.

I am using the same logic in AIO and it works great over there.
https://github.com/nextcloud/all-in-one/blob/feffba739ae5ef2911782514fa5de367d52cf811/php/src/Data/ConfigurationManager.php#L455-L458

### CODE REVIEW

File: lib/private/Config.php [hunk 1]
```diff
@@ -290,6 +290,12 @@ private function writeData() {
 			throw new \Exception(sprintf('Could not acquire an exclusive lock on the config file %s', $this->configFilePath));
 		}
 
+		// Never write file back if disk space should be low (less than 10 MB)
+		$df = disk_free_space($this->configDir);
+		if ($df !== false && (int)$df < 10240) {
```
Comment (st3iny) [inline lib/private/Config.php:295]:
```suggestion
		if ($df !== false && (int)$df < 10240000) {
```

The function returns free disk space in bytes.

Ref https://www.php.net/manual/en/function.disk-free-space.php#refsect1-function.disk-free-space-returnvalues

Comment (szaimen) [reply lib/private/Config.php:295]: ah you are right. I made a mistake in the comment. it should not be less than 10KB but we could think about increasing to 100KB. WDYT?

Comment (szaimen) [reply lib/private/Config.php]: fixed

**Review by come-nc — APPROVED** [review]

File: lib/private/Config.php [hunk 2]
```diff
@@ -290,6 +290,12 @@ private function writeData() {
 			throw new \Exception(sprintf('Could not acquire an exclusive lock on the config file %s', $this->configFilePath));
 		}
 
+		// Never write file back if disk space should be low (less than 100 KiB)
+		$df = disk_free_space($this->configDir);
+		if ($df !== false && (int)$df < 102400) {
+			throw new \Exception($this->configDir . " does not have enough space for writing the config file! Not writing it back!");
+		}
```
Comment (szaimen) [inline lib/private/Config.php:297]: maybe we should move this over the flock?

Comment (szaimen) [reply lib/private/Config.php:297]: edit: done

File: lib/private/Config.php [hunk 3]
```diff
@@ -285,6 +285,12 @@ private function writeData() {
 				'This can usually be fixed by giving the webserver write access to the config directory.');
 		}
 
+		// Never write file back if disk space should be low (less than 100 KiB)
+		$df = disk_free_space($this->configDir);
+		if ($df !== false && (int)$df < 102400) {
```
Comment (nickvergessen) [inline lib/private/Config.php:290]: Should we check against `strlen($content)`? Just in case with many LDAPs, external storages, etc we don't need more?

Comment (szaimen) [reply lib/private/Config.php:290]: fine by me

Comment (szaimen) [reply lib/private/Config.php]: done. Plese check!

**Review by mejo- — APPROVED** [review]

**Comment by szaimen** [comment]
CI failure unrelated

**Comment by MichaIng** [comment]
At least on my 32-bit instance this breaks the update. I just removed `(int)` and now it works. Is there a way to allow it being an int as well as a float without removing the hint completely?

**Pull request merged**